File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree-iii/submission-0.js
/**
 * // Definition for a Node.
 * function Node(val) {
 *    this.val = val;
 *    this.left = null;
 *    this.right = null;
 *    this.parent = null;
 * }
 */
class Solution {
    /**
     * @param {Node} p
     * @param {Node} q
     * @return {Node} 
     * 3-5, 5-12-1, 1-3, 3-5, 5-12-1, 3-5
     * 12-1, 1-3, 3-5, 5-12-1, 1-3, 3-5
     */
    lowestCommonAncestor(p, q) {
        const phead = p;
        const qhead = q;
        while (p !== q) {
            if (p.parent === null) {
                p = qhead;
            } else {
                p = p.parent;
            }
            if (q.parent === null) {
                q = phead;
            } else {
                q = q.parent;
            }
            
        }
        return p; // both nodes now point at LCA
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # simple, brute force is to count all nodes, build array of all nodes in sorted order
        nodes = []
        def dfs(root) -> None:
            if root == None:
                return

            nonlocal nodes
            if root.left:
                dfs(root.left)
            nodes.append(root.val)
            if root.right:
                dfs(root.right)
        dfs(root)

        # now that nodes has all vals in sorted order, step through k
        return nodes[k-1]